### ISSUE TITLE
Ascertain whether it's possible to hook into and override theme.json settings in a particular editor context

### ISSUE THREAD

**Issue by getdave** [WordPress/gutenberg#34775]
Part of https://github.com/WordPress/gutenberg/issues/30007#issuecomment-893575752

> Ascertain whether it's possible to hook into and override theme.json settings in a particular editor context (e.g. can we override certain settings within the Nav Editor

Please see the link above for _why_ we might we want to do this.

Currently the global styles settings are derived from x3 sources

1. Core
2. Theme (`theme.json`)
3. User (global styles UI - incomplete).

There is [a resolver which collates these disparate sources and merges them to form the final global styles settings](https://github.com/WordPress/gutenberg/blob/de446ff2008c82b90110a329d9109a8e85bede3e/lib/class-wp-theme-json-resolver-gutenberg.php#L409).

Currently there is - by design or otherwise - no way to filter these settings.

## Potential Solution 
We could consider proposing a filter here:

https://github.com/WordPress/gutenberg/blob/de446ff2008c82b90110a329d9109a8e85bede3e/lib/class-wp-theme-json-resolver-gutenberg.php#L427

It might look like this:

```diff
+ $result = apply_filters( 'theme_json_resolver_merged_data', $result );

return $result;
```

This would allow the Nav Editor to filter the settings.

## Considerations

Is it a good idea to allow developers to potentially override User settings (recall User is one of the x3 sources for Global Styles settings)?

We could consider only allow overriding of Core/Theme settings. That would leave users fully "in charge" but (hopefully) still allow us to filter the settings in the Nav Editor. If the Nav _block_ was configured via Theme JSON then this would allow us to alter the behaviour/display of the Nav block whilst used in the Nav Editor.

**Comment by sdwire** [comment]
Here's another reason one might want to hook in and override theme.json settings. I'm hosting a multi-site installation. In that installation, I offer a custom but highly configurable theme. Different sites using the same theme are configured differently. I would want my theme to have a chance to make some of those theme.json-configured decisions differently for different sites, depending on their dynamic configuration. In my case, the different editor contexts would be different subsites using the same installation of the same theme, but with different site-specific configurations.

For example, different sites may have different blocks available, or have certain capabilities of certain blocks enabled or disabled depending on the site-specific configuration.

I have been searching for how to make that happen, and this issue suggests to me that it's not (yet) possible. I do like the proposed idea for a filter that would allow the theme to have one last chance to update the merged configuration before the block editor responds to it.

If there's another way for the theme programming to make some of the theme.json decisions dynamically, I'd welcome that guidance as well.

**Comment by getdave** [comment]
You are right. I explored this and it's not possible...yet. I'm no longer actively pursuing this route but it's possible it might be something that is considered in future.

**Comment by tyxla** [comment]
I'm closing this one as it's not relevant. 

WordPress now has hooks for this:

- `wp_theme_json_data_default`, `wp_theme_json_data_blocks`, `wp_theme_json_data_theme` and `wp_theme_json_data_user` let you change theme.json data. They receive a `WP_Theme_JSON_Data` object, not an editor context.
- `block_editor_settings_all` receives the editor context, so use it for settings in one editor only. It runs after the global settings are merged. Some older settings are built from them, so check those too. Changing editor settings does not change the frontend CSS.

The `theme_json_resolver_merged_data` filter was never added. 

If the above hooks don't cover your case, feel free to share the exact use case, and we can reopen.
